invoke.go

```diff
@@ -17,21 +17,28 @@ func scaleDuration(a time.Duration, mult float64) time.Duration {
 	return time.Duration(ns)
 }
 
+// contextWithTimeout returns a context with the given timeout applied if there
+// is no deadline on the context.
+func contextWithTimeout(ctx context.Context, timeout time.Duration) context.Context {
```

Comment:
nit: a name that doesn't suggest context.WithTimeout. How about "ensureTimeout" since it ensures that the input context will have some timeout?